### PR TITLE
[Legacy] Add AddTopIncludeRector

### PR DESCRIPTION
**Pull request by phpfui**
This will add an include type statement to any plain PHP file, but not class definitions.  It does add a new include statement each time it is run, but I think that is OK, as you may want to add more than one (but not sure why). Can be combined with the RemoveIncludeRector to nuke the old includes.

### CODE REVIEW

File: rules/legacy/src/Rector/Include_/AddTopIncludeRector.php [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+    }
+
+    /**
+     * @return string[]
+     */
+    public function getNodeTypes(): array
+    {
+        $this->initialize();
+
+        return [Node::class];
```
Comment (TomasVotruba) [inline rules/legacy/src/Rector/Include_/AddTopIncludeRector.php:115]: If you want to use somethign like `Node`, that's a sing this should be a `AbstractFileSystemRector`

Comment (phpfui) [reply rules/legacy/src/Rector/Include_/AddTopIncludeRector.php:115]: So finally had some time to look at this.  If I were to make it a AbstractFileSystemRector, then I would have to use BetterNodeFinder.  I can find a class node with that, but I also want to find the first executable statement and insert the include before that.  I currently just filter out nodes that are non-executable, then insert when I find a node that is not in that list. Not sure how I would do that using BetterNodeFinder.  There are some filters available, but not sure it is any better than what I am doing now.

File: rules/legacy/src/Rector/Include_/AddTopIncludeRector.php [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+if (isset($_POST["csrf"])) {
+    processPost($_POST);
+}',
+                '<?php
+
+include "autoloader.php";
+
+if (isset($_POST["csrf"])) {
+    processPost($_POST);
+}',
```
Comment (TomasVotruba) [inline rules/legacy/src/Rector/Include_/AddTopIncludeRector.php:96]: In multine-samples, there should be now/heredoc format used: 
https://github.com/rectorphp/rector/blob/965cef98e218163cb43049c1e5ef0b32fa635eef/rules/renaming/src/Rector/Class_/RenameClassRector.php#L55-L79

This prevents spacing issues and missing newlines

File: rules/legacy/src/Rector/Include_/AddTopIncludeRector.php (hunk 1 above)
Comment (TomasVotruba) [reply rules/legacy/src/Rector/Include_/AddTopIncludeRector.php:115]: There is no need to use `BetterNodeFinder`, if you want to update top-level statements.

`foreach` over of nodes should do.

Could you update test fixtures with ~2 more samples that will cover edge-cases you discovered? 

Also 1 fixture with `skip_xxx.php.inc`, that checks the include is already there and to prevent double adding it.


<br>

I'll prepare short code sample for `AbstractFileSystemRector`, so it's easier to understand

File: rules/legacy/src/Rector/Include_/AddTopIncludeRector.php [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+            $this->completed = false;
+        }
+
+        if (! $this->matchFile($path)) {
+            $this->completed = true;
+
+            return null;
+        }
+
+        $name = get_class($node);
```
Comment (TomasVotruba) [inline rules/legacy/src/Rector/Include_/AddTopIncludeRector.php:134]:
```suggestion
        $nodeClass = get_class($node);
```

File: rules/legacy/src/Rector/Include_/AddTopIncludeRector.php [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+        $this->addNodeBeforeNode($includeNode, $node);
+        $this->addNodeBeforeNode(new Nop(), $node);
+
+        return $node;
+    }
+
+    /**
+     * Match file against matches, no patterns provided, then it matches
+     */
+    private function matchFile(string $path): bool
```
Comment (TomasVotruba) [inline rules/legacy/src/Rector/Include_/AddTopIncludeRector.php:161]: :+1:

File: rules/legacy/src/Rector/Include_/AddTopIncludeRector.php (hunk 1 above)
Comment (TomasVotruba) [reply rules/legacy/src/Rector/Include_/AddTopIncludeRector.php:115]: Here is a good example:

https://github.com/rectorphp/rector/blob/965cef98e218163cb43049c1e5ef0b32fa635eef/rules/legacy/src/Rector/FileSystem/FunctionToStaticMethodRector.php#L70-L73



```php
    public function refactor(SmartFileInfo $smartFileInfo): void
    {
        $nodes = $this->parseFileInfoToNodes($smartFileInfo);
        $fileStmts = $this->getFileOrNamespaceStmts($nodes);

        // 0. has already include? → skip
        foreach ($fileStmts as $fileStmt) {
           if ($this->isTopFileInclude(fileStmt)) {
               return;
           }
		}
      
        // 1. add include to right location
        foreach ($fileStmts as $fileStmt) {
           if (! $this->isLocationToAddInclude(fileStmt)) {
               continue;
           }
             
           // add it
           $fileStmts = ...;
		}

        // 2. print stmts back
        // ... mimic the process from the rule
   }
```

To add a stmt betweent other statements, you can use `array_splice`, but that was too magic for me and very hard to read/debug. So I use simple foreach:

```php
    $newFileStmts = [];
    foreach ($fileStmts as $fileStmt) {
        if ($this->isLocationToAddInclude(fileStmt)) {
           // add it
           $newFileStmts[] = new Expression(new Include_(...));               continue;
        }

        $newFileStmts[] = $fileStmt;
    }


// $this->print...( $newFileStmts)
```

File: rules/legacy/src/Rector/Include_/AddTopIncludeRector.php (hunk 3 above)
Comment (phpfui) [reply rules/legacy/src/Rector/Include_/AddTopIncludeRector.php:134]: So you are saying if I move to AbstractFileSystemRector, I can look for class and include (of the same file name), and if neither is found, I can insert before the first node and that should work?

I'll will check to see if that would handle InlineHTML before the first executable statement.  That is the only exception that might be an issue.

File: rules/legacy/src/Rector/Include_/AddTopIncludeRector.php (hunk 1 above)
Comment (phpfui) [reply rules/legacy/src/Rector/Include_/AddTopIncludeRector.php:115]: OK, thanks for the example!  They really help to understand the code.

File: rules/legacy/src/Rector/Include_/AddTopIncludeRector.php (hunk 3 above)
Comment (TomasVotruba) [reply rules/legacy/src/Rector/Include_/AddTopIncludeRector.php:134]: > I can look for class and include (of the same file name), and if neither is found, I can insert before the first node and that should work?

Partially. It should also work with other includes of different files, e.g.

```diff
+include __DIR__ . '/autoloader.php' # be sure the paths is absolute, to prevent magic with relative paths 

 include 'some_other_file.php;

 if (...) {
    include 'yet_another.php;
 }
```

File: rules/legacy/src/Rector/Include_/AddTopIncludeRector.php (hunk 1 above)
Comment (TomasVotruba) [reply rules/legacy/src/Rector/Include_/AddTopIncludeRector.php:115]: If that's too much to absorb now and you'll feel stressed, just add the test fixture and I'll merge it and do the change myself for this 1st case.

Comment (phpfui) [reply rules/legacy/src/Rector/Include_/AddTopIncludeRector.php:115]: Let me take another whack at it.  I think I understand what needs to be done.  The biggest issue for me is the PHPStan checks don't seem to run well under Windows 10. They also seem to reconfigure the method single line into multiple lines, which fail on github.  I prefer the single line format myself, so I was confused when it multi lined them.

I was also wondering about the string expression that could be in a include statement.  How would I specify something like __DIR__ . '/../file.php' in the config and get it converted into nodes?

Comment (phpfui) [reply rules/legacy/src/Rector/Include_/AddTopIncludeRector.php:115]: Hmm, I could probably just send the string through PhpParser, then add that expression to the include node.

Comment (TomasVotruba) [reply rules/legacy/src/Rector/Include_/AddTopIncludeRector.php:115]: @phpfui 

> Hmm, I could probably just send the string through PhpParser, then add that expression to the include node.

Exactly! :+1:

Comment (TomasVotruba) [reply rules/legacy/src/Rector/Include_/AddTopIncludeRector.php:115]: > The biggest issue for me is the PHPStan checks don't seem to run well under Windows 10.

I'm sorry to hear that. I actually don't know many CLI tools that work on Windows. It's often quite risky. I used Windows for PHP, but I do use Ubuntu for last 10 years and it's much easier as it just works. You might find more troubles with Rector on Windows in the future, just so you're aware of it.

**Comment by phpfui** [comment]
I figured out how to call PHPStan from Windows.  Next commit will have that fixed.

Question is how do I save the nodes in this mode?  I called ~~~$this->print($nodes)~~ but that does not seem to work.

**Comment by TomasVotruba** [comment]
> Question is how do I save the nodes in this mode? I called ~~~$this->print($nodes)~~ but that does not seem to work.

If you check the other rules that extends `AbstratcFileSystemRector`, there should be a method for that

File: rules/legacy/src/Rector/Include_/AddTopIncludeRector.php [hunk 5]
```diff
@@ -181,20 +154,48 @@ private function initialize(): void
...
+        // normalize the file we are including
+        $this->file = $this->getStringFromExpression($this->fileExpression);
+        if (isset($this->settings['match']) && is_array($this->settings['match'])) {
+            $this->patterns = $this->settings['match'];
+        }
+    }
+
+    private function getExpressionFromString(string $source): ?Expr
+    {
+        $ast = $this->parser->parse("<?php {$source};\n");
```
Comment (TomasVotruba) [inline rules/legacy/src/Rector/Include_/AddTopIncludeRector.php:179]: Better use `$nodes` convention, so we have the same naming in all files

Comment (TomasVotruba) [reply rules/legacy/src/Rector/Include_/AddTopIncludeRector.php:179]: Also, the "in string {$source}" should be prevented too.

```suggestion
        $nodes = $this->parser->parse('<?php ' . $source . ';');
```

 It makes PHPStan + Rector refactoring more complicated.


PHPStan missed that?

File: rules/legacy/src/Rector/Include_/AddTopIncludeRector.php [hunk 6]
```diff
@@ -181,20 +154,48 @@ private function initialize(): void
...
+        if (! isset($constants[$this->settings['type'] ?? ''])) {
+            throw new InvalidRectorsettingsException('Invalid type: must be one of ' . implode(
                 ', ',
                 array_keys($constants)
             ));
         }
 
-        if ('' === ($this->configuration['file'] ?? '')) {
-            throw new InvalidRectorConfigurationException('Invalid parameter: file must be provided');
+        if ('' === ($this->settings['file'] ?? '')) {
```
Comment (TomasVotruba) [inline rules/legacy/src/Rector/Include_/AddTopIncludeRector.php:164]: 3 operations at one line. This is bit confusing to read... what is the goal of this ?

My guess: 

```suggestion
        if (! isset($this->settings['file'])) {
```

Comment (phpfui) [reply rules/legacy/src/Rector/Include_/AddTopIncludeRector.php:164]: I originally had empty($this->settings['file']) but PHPStan complained "Use strict comparison instead of empty".  empty is actually correct in this situation. If the index is not set, it is an error and empty.  If the index is set, but empty ('', 0, etc), it is also an error.  It needs to be a string of some sort.  isset would need an additional test.

So I went with the null coalesce operator, but you have to wrap it in () to make sure the operator precedence is always correct, then type compare it to string to make sure it is not empty.  I could reverse the lhs and rhs, but it is my style to put the scalar constant on the lhs so if you mistype a comparison as as assign, you get an error. I have noticed your style guide does not do this.  I would recommend it.

Other suggestions?

Comment (TomasVotruba) [reply rules/legacy/src/Rector/Include_/AddTopIncludeRector.php:158]: I wonder how to make it simple to prevent such a rocket engeneering over one variable.

What about makint it a parameter? The file name is required anyway, right?

File: rules/legacy/tests/Rector/Include_/AddTopIncludeRector/AddTopIncludeRectorTest.php [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+
+    protected function getRectorsWithConfiguration(): array
+    {
+        return [
+            AddTopIncludeRector::class => [
+                '$settings' => [
+                    'type' => 'TYPE_INCLUDE',
+                    'file' => '__DIR__ . "/../autoloader.php"',
+                ],
+            ],
```
Comment (TomasVotruba) [inline rules/legacy/tests/Rector/Include_/AddTopIncludeRector/AddTopIncludeRectorTest.php:35]: Ah, I see it now. Array makes sense only in case of multiple instances, like rename these 30 classes to these 30 classes.

With this 1-level setup, you can use them directly:

```suggestion
            AddTopIncludeRector::class => [
                '$type' => 'TYPE_INCLUDE',
                '$file' => '__DIR__ . "/../autoloader.php"',
            ],
```

File: rules/legacy/src/Rector/Include_/AddTopIncludeRector.php (hunk 6 above)
Comment (TomasVotruba) [reply rules/legacy/src/Rector/Include_/AddTopIncludeRector.php:158]: See: https://github.com/rectorphp/rector/pull/3688#discussion_r454419900

File: rules/legacy/tests/Rector/Include_/AddTopIncludeRector/AddTopIncludeRectorTest.php (hunk 7 above)
Comment (phpfui) [reply rules/legacy/tests/Rector/Include_/AddTopIncludeRector/AddTopIncludeRectorTest.php:35]: OK, that is what I was trying to figure out, but could not seem to get it to work.  Will change.

**Comment by TomasVotruba** [comment]
I guess you've tried that and it fails on Windows... but if not, you can fix all coding standard issues before commiting like this:

```bash
composer fix-cs
```

**Comment by phpfui** [comment]
fix-cs does not find all issues, and makes changes that if checked in will
not pass the Linux version, like multiline method declarations.  Very
strange.

Seems to run under Windows though, so not sure why it is finding and fixing
different things.

**Comment by TomasVotruba** [comment]
Don't worry about it :+1: . I think it's ready now.  

Could you squash it to 1 commit?

I'll merge it then and review locally

**Comment by phpfui** [comment]
Yes, I finally think it is good. Squash coming up.

I did notice that PHPUnit seems to think the tests are risky with no
assertions made.  Not sure why that is, you may want to look into it and
fix.  I just don't know enough about the unit tests to fix.  I think I got
it right, but obviously not.

**Comment by phpfui** [comment]
Closing.  Opening squashed PR.
